File: backend/app/models/academic_calendar.py
"""Academic calendar event model for semester milestones, exam periods, holidays, etc."""

from datetime import datetime
from typing import Optional
from enum import Enum

from pydantic import BaseModel, Field


class AcademicEventType(str, Enum):
    """Types of academic calendar events."""
    exam_period = "exam_period"
    registration = "registration"
    add_drop = "add_drop"
    holiday = "holiday"
    break_period = "break_period"
    orientation = "orientation"
    convocation = "convocation"
    lecture_start = "lecture_start"
    lecture_end = "lecture_end"
    deadline = "deadline"
    other = "other"


EVENT_TYPE_LABELS = {
    AcademicEventType.exam_period: "Exam Period",
    AcademicEventType.registration: "Course Registration",
    AcademicEventType.add_drop: "Add/Drop Period",
    AcademicEventType.holiday: "Holiday",
    AcademicEventType.break_period: "Break",
    AcademicEventType.orientation: "Orientation",
    AcademicEventType.convocation: "Convocation",
    AcademicEventType.lecture_start: "Lectures Begin",
    AcademicEventType.lecture_end: "Lectures End",
    AcademicEventType.deadline: "Deadline",
    AcademicEventType.other: "Other",
}


class AcademicEventCreate(BaseModel):
    """Schema for creating an academic calendar event."""
    title: str = Field(..., min_length=2, max_length=200)
    eventType: AcademicEventType
    startDate: datetime
    endDate: Optional[datetime] = None
    semester: int = Field(..., ge=1, le=2)
    description: Optional[str] = Field(None, max_length=500)


class AcademicEventUpdate(BaseModel):
    """Schema for updating an academic calendar event."""
    title: Optional[str] = Field(None, min_length=2, max_length=200)
    eventType: Optional[AcademicEventType] = None
    startDate: Optional[datetime] = None
    endDate: Optional[datetime] = None
    semester: Optional[int] = Field(None, ge=1, le=2)
    description: Optional[str] = Field(None, max_length=500)


class AcademicEventResponse(BaseModel):
    """Schema for academic calendar event response."""
    id: str
    title: str
    eventType: AcademicEventType
    startDate: datetime
    endDate: Optional[datetime] = None
    semester: int
    sessionId: str
    description: Optional[str] = None
    createdBy: str
    createdAt: datetime
    updatedAt: Optional[datetime] = None

File: src/lib/api/academic-calendar.ts
/**
 * Academic Calendar API service — types, constants, and API functions
 * for managing semester milestones, exam periods, holidays, etc.
 */

import { api, buildQueryString } from "./client";

// ── Types ────────────────────────────────────────────────────────────────

export type AcademicEventType =
  | "exam_period"
  | "registration"
  | "add_drop"
  | "holiday"
  | "break_period"
  | "orientation"
  | "convocation"
  | "lecture_start"
  | "lecture_end"
  | "deadline"
  | "other";

export interface AcademicEvent {
  id: string;
  title: string;
  eventType: AcademicEventType;
  startDate: string;
  endDate?: string | null;
  semester: number;
  sessionId: string;
  description?: string | null;
  createdBy: string;
  createdAt: string;
  updatedAt?: string | null;
}

export interface CreateAcademicEventData {
  title: string;
  eventType: AcademicEventType;
  startDate: string;
  endDate?: string | null;
  semester: number;
  description?: string;
}

export interface UpdateAcademicEventData {
  title?: string;
  eventType?: AcademicEventType;
  startDate?: string;
  endDate?: string | null;
  semester?: number;
  description?: string;
}

// ── Constants ────────────────────────────────────────────────────────────

export const EVENT_TYPE_LABELS: Record<AcademicEventType, string> = {
  exam_period: "Exam Period",
  registration: "Course Registration",
  add_drop: "Add/Drop Period",
  holiday: "Holiday",
  break_period: "Break",
  orientation: "Orientation",
  convocation: "Convocation",
  lecture_start: "Lectures Begin",
  lecture_end: "Lectures End",
  deadline: "Deadline",
  other: "Other",
};

export const EVENT_TYPE_COLORS: Record<
  AcademicEventType,
  { bg: string; text: string; border: string }
> = {
  exam_period: { bg: "bg-coral-light", text: "text-coral", border: "border-coral" },
  registration: { bg: "bg-teal-light", text: "text-teal", border: "border-teal" },
  add_drop: { bg: "bg-sunny-light", text: "text-navy", border: "border-sunny" },
  holiday: { bg: "bg-lavender-light", text: "text-lavender", border: "border-lavender" },
  break_period: { bg: "bg-lavender-light", text: "text-lavender", border: "border-lavender" },
  orientation: { bg: "bg-teal-light", text: "text-teal", border: "border-teal" },
  convocation: { bg: "bg-lime-light", text: "text-navy", border: "border-lime" },
  lecture_start: { bg: "bg-teal-light", text: "text-teal", border: "border-teal" },
  lecture_end: { bg: "bg-coral-light", text: "text-coral", border: "border-coral" },
  deadline: { bg: "bg-sunny-light", text: "text-navy", border: "border-sunny" },
  other: { bg: "bg-ghost", text: "text-navy", border: "border-cloud" },
};

export const ALL_EVENT_TYPES: AcademicEventType[] = [
  "exam_period",
  "registration",
  "add_drop",
  "holiday",
  "break_period",
  "orientation",
  "convocation",
  "lecture_start",
  "lecture_end",
  "deadline",
  "other",
];

// ── API Functions ────────────────────────────────────────────────────────

const BASE = "/api/v1/academic-calendar";

export async function listAcademicEvents(filters?: {
  semester?: number;
  eventType?: string;
}): Promise<AcademicEvent[]> {
  const qs = buildQueryString(filters || {});
  return api.get<AcademicEvent[]>(`${BASE}/${qs}`);
}

export async function getAcademicEvent(id: string): Promise<AcademicEvent> {
  return api.get<AcademicEvent>(`${BASE}/${id}`);
}

export async function createAcademicEvent(
  data: CreateAcademicEventData
): Promise<AcademicEvent> {
  return api.post<AcademicEvent>(BASE + "/", data);
}

export async function updateAcademicEvent(
  id: string,
  data: UpdateAcademicEventData
): Promise<AcademicEvent> {
  return api.patch<AcademicEvent>(`${BASE}/${id}`, data);
}

export async function deleteAcademicEvent(id: string): Promise<void> {
  return api.delete(`${BASE}/${id}`);
}
